### PR TITLE
JIT: Have physical promotion insert read backs before possible implicit control flow

### LINKED ISSUE #86498: JIT: Physical promotion readback mechanism does not take implicit control flow into account
Physical promotion currently relies on being able to insert pending read backs before control is transferred to any successor block. This happens at the end of the block, but doing it this way does not take implicit control flow into account. For example:

```csharp
using System;
using System.Runtime.CompilerServices;

public class Program
{
    public static int Main()
    {
        int result = Test();
        if (result == 100)
        {
            Console.WriteLine("PASS");
        }
        else
        {
            Console.WriteLine("FAIL: Returned {0}", result);
        }

        return result;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int Test()
    {
        Foo f = new();
        try
        {
            f.X = 15;
            f.Y = 20;
            f.X += f.Y;
            f.Y *= f.X;

            // f will be physically promoted and will require a read back after this call.
            // However, there is implicit control flow happening that the read back should happen before.
            f = Call(f);
            ThrowException();
            return -1;
        }
        catch (Exception ex)
        {
            return f.X + f.Y;
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static Foo Call(Foo f)
    {
        return new Foo { X = 75, Y = 25 };
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void ThrowException()
    {
        throw new Exception();
    }

    private struct Foo
    {
        public short X, Y;
    }
}
```

To fix we can likely insert any pending read backs before any `GTF_EXCEPT` marked tree when in a try region.

### PR DESCRIPTION
**Pull request by jakobbotsch**
Physical promotion relies on being able to read back any promoted field that is fresher in its struct local before control flows to any successor block. This was failing to take implicit control flow into account.

Fix #86498

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS 

Small diffs. [With physical promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=284760&view=ms.vss-build-web.run-extensions-tab), [without old promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=284761&view=ms.vss-build-web.run-extensions-tab).

File: src/coreclr/jit/promotion.cpp [hunk 1]
```diff
@@ -1449,11 +1583,11 @@ void ReplaceVisitor::WriteBackBefore(GenTree** use, unsigned lcl, unsigned offs,
 //   offs         - The starting offset of the range in the struct local that needs to be read back from.
 //   size         - The size of the range
 //
-void ReplaceVisitor::MarkForReadBack(unsigned lcl, unsigned offs, unsigned size)
+bool ReplaceVisitor::MarkForReadBack(unsigned lcl, unsigned offs, unsigned size)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/promotion.cpp:1586]: Document the return value?

File: src/coreclr/jit/promotion.cpp [hunk 2]
```diff
@@ -1112,10 +1112,68 @@ GenTree* Promotion::CreateReadBack(Compiler* compiler, unsigned structLclNum, co
...
+            continue;
+        }
+
+        for (size_t i = 0; i < agg->Replacements.size(); i++)
+        {
+            Replacement& rep = agg->Replacements[i];
+            assert(!rep.NeedsReadBack || !rep.NeedsWriteBack);
+            if (rep.NeedsReadBack)
+            {
+                if (m_liveness->IsReplacementLiveOut(m_currentBlock, agg->LclNum, (unsigned)i))
```
Comment (AndyAyersMS) [inline src/coreclr/jit/promotion.cpp:1143]: I suppose this is where you get into the resolution strategies... if the field is partially dead (only live in some successors) you might be tempted to defer the read back until the start of the live successors (if join free, or if all other preds of those successors will do read backs too).

We might be able to get at some of this already via tail merging.

File: src/coreclr/jit/promotion.cpp [hunk 3]
```diff
@@ -1112,10 +1112,68 @@ GenTree* Promotion::CreateReadBack(Compiler* compiler, unsigned structLclNum, co
...
+    }
+
+    m_hasPendingReadBacks = false;
+}
+
 Compiler::fgWalkResult ReplaceVisitor::PostOrderVisit(GenTree** use, GenTree* user)
 {
     GenTree* tree = *use;
 
+    use = InsertMidTreeReadBacksIfNecessary(use);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/promotion.cpp:1175]: Add a header comment for PostOrderVisit.

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/promotion.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/promotion.cpp:1586]: In my next change I remove this return value in favor of moving the logging into this function, so will address this with that PR.

File: src/coreclr/jit/promotion.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/promotion.cpp:1143]: Yeah, I think there's a few opportunities like that around. I liked your idea of focusing on a linear trace when making these decisions.

File: src/coreclr/jit/promotion.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/promotion.cpp:1175]: Will add it as part of my follow-up.

**Pull request merged**